application_sandbox/copy_querypool_results_host_reset/main.cpp

```diff
@@ -0,0 +1,525 @@
+                                       .EnableDepthBuffer()
+                                       .EnableMultisampling()
+                                       .EnableHostQueryReset(),
+                                   requested_features,
+                                   {"VK_KHR_get_physical_device_properties2"},
+                                   {VK_EXT_HOST_QUERY_RESET_EXTENSION_NAME}),
+        torus_(data->allocator(), data->logger(), torus_data),
+        grey_scale_(0u),
+        num_frames_(0u) {}
+  virtual void InitializeApplicationData(
+      vulkan::VkCommandBuffer* initialization_buffer,
+      size_t num_swapchain_images) override {
+    num_frames_ = static_cast<uint32_t>(num_swapchain_images);
+
+    // For GAPID, when Mid-Execution Capture is applied, we need to reconstruct
+    // the graphics state before starting tracing the following graphics
+    // commands.  However, the query results are dependent on the draw calls
+    // before the reconstruction of the graphics state so they cannot be
+    // reconstructed completely. This means the first several query pool
+    // results obtained on the replay device can be incorrect when
+    // Mid-Execution Capture is turned on.
+    // This sample obtains the query pool results prior to the draw call for
+    // each frame, and uses the query results to determine the greyscale of the
+    // model to be drawn. In a normal run, the first draw call of each frame
+    // will print a blank (just black screen), and the following ones will show
+    // the model. When Mid-Execution Capture is applied, no matter on which
+    // frame the tracing is started, we expect the first draw call of each
+    // frame on the replay device will always be blank, and the following ones
+    // will show the model.
+    // E.g.: Mid-Execution Capture starting on Frame 100, Number of swapchain
+    // images on tracing and replay device: 3.
+    // Tracing device: First 3 frames are blank. Frame 100, 101, 102 show the
+    // model.
+    // Replay device: First 3 frames (corresponding to frame 100, 101, 102 on
+    // the tracing device) are blank, following ones shows the model.
+
+    // Create a occlusion query pool that contains a query for each frame.
+    VkQueryPoolCreateInfo query_pool_create_info{
+        VK_STRUCTURE_TYPE_QUERY_POOL_CREATE_INFO,  // sType
+        nullptr,                                   // pNext
+        0,
+        VK_QUERY_TYPE_OCCLUSION,
+        uint32_t(num_swapchain_images),
+        0,
+    };
+    query_pool_ = containers::make_unique<vulkan::VkQueryPool>(
+        data_->allocator(),
+        vulkan::CreateQueryPool(&app()->device(), query_pool_create_info));
+
+    // Query before drawing anything to make sure the initial value of query
+    // pool results are zero.
+    for (size_t i = 0; i < num_swapchain_images; i++) {
+      app()->device()->vkResetQueryPoolEXT(app()->device(), *query_pool_,
```

Comment:
Should this be a host reset too?

Reply:
I don't think this particular reset matters. It happens only once to guarantee the query pool is reset initially but we are resetting it anyway before each frame (before the CB is submitted). So it doesn't really matter if it's a host reset or a device reset.

Reply:
Might as well just replace all resets with the host version?  TBH I don't think it matters, just an idea.